aws/resource_aws_appsync_api_key_test.go

```diff
@@ -0,0 +1,114 @@
+package aws
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/appsync"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccAwsAppsyncApiKey_basic(t *testing.T) {
+	// sample date to test
+	dateAfterOneYear := time.Now().Add(time.Hour * 24 * time.Duration(364))
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAwsAppsyncApiKeyDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config:       testAccAppsyncApiKeyConfigBasic(dateAfterOneYear.Format(time.RFC3339)),
+				ResourceName: "aws_appsync_api_key.test",
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsAppsyncApiKeyExistsTillDate(
+						"aws_appsync_api_key.test",
+						dateAfterOneYear.Unix(),
+					),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckAwsAppsyncApiKeyDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).appsyncconn
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_appsync_api_key" {
+			continue
+		}
+		ApiId, _, er := decodeAppSyncApiKeyId(rs.Primary.ID)
+		if er != nil {
+			return er
+		}
+		describe, err := conn.ListApiKeys(&appsync.ListApiKeysInput{ApiId: aws.String(ApiId)})
+		if err == nil {
+			if len(describe.ApiKeys) != 0 {
+				return fmt.Errorf("Appsync ApiKey still exists")
+			}
+		}
+
+		return nil
+
+	}
+	return nil
+}
+
+func testAccCheckAwsAppsyncApiKeyExistsTillDate(ApiKeyName string, date int64) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+
+		rsApiKey, ok := s.RootModule().Resources[ApiKeyName]
+		if !ok {
+			return fmt.Errorf("Key Not found in state: %s", ApiKeyName)
+		}
+		ApiId, Id, er := decodeAppSyncApiKeyId(rsApiKey.Primary.ID)
+		if er != nil {
+			return er
+		}
+
+		conn := testAccProvider.Meta().(*AWSClient).appsyncconn
+		input := &appsync.ListApiKeysInput{
+			ApiId: aws.String(ApiId),
+		}
+
+		resp, err := conn.ListApiKeys(input)
+		if err != nil {
+			return err
+		}
+
+		var key appsync.ApiKey
+		for _, v := range resp.ApiKeys {
+			if *v.Id == *aws.String(Id) {
+				key = *v
```

Comment:
I would recommend splitting this logic into two functions:

```go
func testAccCheckAwsAppsyncApiKeyExists(resourceName string, apiKey *appsync.ApiKey)
```

which calls `*apiKey = *v` right here or otherwise returns an error that the key could not be found.

Then separately:

```go
func testAccCheckAwsAppsyncApiKeyTillDate(apiKey *appsync.ApiKey, date time.Time)
```

which handles the date/time logic below.